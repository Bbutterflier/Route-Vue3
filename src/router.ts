import { createRouter, createWebHashHistory, RouteParams } from 'vue-router'

import Home from './pages/Home.vue'

export type AppRouteNames =
  | 'global-feed'
  | 'feed'
  | 'login'
  | 'tag'
  | 'register'
  | 'settings'
  | 'profile'
  | 'editor'

export const router = createRouter({
  history: createWebHashHistory(),
  routes: [
    {
      name: 'global-feed',
      path: '/',
      component: Home,
    },
    {
      name: 'feed',
      path: '/',
      component: Home,
    },
    {
      name: 'tag',
      path: '/tag/:tag',
      component: Home,
    },
    {
      name: 'login',
      path: '/login',
      component: () => import('./pages/Login.vue'),
    },
    {
      name: 'register',
      path: '/register',
      component: () => import('./pages/Register.vue'),
    },
    {
      name: 'settings',
      path: '/settings',
      component: () => import('./pages/Settings.vue'),
    },
    {
      name: 'profile',
      path: '/settings',
      component: Home,
    },
    {
      name: 'editor',
      path: '/settings',
      component: Home,
    },
  ],
})

export function routerPush(
  name: AppRouteNames,
  params?: RouteParams
): ReturnType<typeof router.push> {
  if (params !== undefined) {
    return router.push({
      name,
      params,
    })
  } else {
    return router.push({
      name,
    })
  }
}
