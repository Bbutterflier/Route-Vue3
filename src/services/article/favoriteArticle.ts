import { request } from '../index'

export function postFavoriteArticle(slug: string): Promise<Article> {
  return request
    .post<ArticleResponse>(`/articles/${slug}/favorite`)
    .then((r) => r.article)
}

export function deleteFavoriteArticle(slug: string): Promise<Article> {
  return request
    .delete<ArticleResponse>(`/articles/${slug}/favorite`)
    .then((r) => r.article)
}
