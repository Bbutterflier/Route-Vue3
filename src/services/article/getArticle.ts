import { request } from '../index'

export function getArticleBySlug(slug: string): Promise<Article> {
  return request
    .get<ArticleResponse>(`/articles/${slug}`)
    .then((r) => r.article)
}
