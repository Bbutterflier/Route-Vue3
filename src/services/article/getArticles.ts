import { limit, request } from '../index'

export function getArticles(page = 1): Promise<ArticlesResponse> {
  const params = { limit, offset: (page - 1) * limit }
  return request.get<ArticlesResponse>('/articles', { params })
}

export function getArticlesByTag(
  tagName: string,
  page = 1
): Promise<ArticlesResponse> {
  const params = { tag: tagName, limit, offset: (page - 1) * limit }
  return request.get<ArticlesResponse>('/articles', { params })
}
