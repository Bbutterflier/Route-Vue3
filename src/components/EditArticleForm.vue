<template>
  <form @submit.prevent="onSubmit">
    <fieldset>
      <fieldset class="form-group">
        <input
          v-model="formData.title"
          type="text"
          class="form-control form-control-lg"
          placeholder="Article Title"
          data-test="title"
        />
      </fieldset>
      <fieldset class="form-group">
        <input
          v-model="formData.description"
          type="text"
          class="form-control"
          placeholder="What's this article about?"
          data-test="description"
        />
      </fieldset>
      <fieldset class="form-group">
        <textarea
          v-model="formData.body"
          class="form-control"
          rows="8"
          placeholder="Write your article (in markdown)"
          data-test="body"
        ></textarea>
      </fieldset>
      <fieldset class="form-group">
        <input
          v-model="newTag"
          type="text"
          class="form-control"
          placeholder="Enter tags"
          @keydown.enter.prevent="addTag"
          data-test="newTag"
        />
        <div class="tag-list" data-test="tag-list">
          <span
            v-for="tag in formData.tagList"
            :key="tag"
            class="tag-default tag-pill"
            @click="removeTag(tag)"
          >
            <i class="ion-close-round" />
            {{ tag }}
          </span>
        </div>
      </fieldset>
      <button class="btn btn-lg pull-xs-right btn-primary" type="submit">
        Publish Article
      </button>
    </fieldset>
  </form>
</template>

<script lang="ts">
export default {
  name: 'EditArticleForm',
}
</script>

<script lang="ts" setup>
import { ref, reactive, computed } from 'vue'
import { useRoute } from 'vue-router'
import { routerPush } from '../router'
import { getArticleBySlug } from '../services/article/getArticle'
import { postArticle, putArticle } from '../services/article/postArticle'

interface Form {
  title: string
  description: string
  body: string
  tagList: string[]
}

const route = useRoute()
const slug = computed(() => route.params.slug as string)

const formData = reactive<Form>({
  title: '',
  description: '',
  body: '',
  tagList: [],
})

const newTag = ref('')

const addTag = () => {
  const val = newTag.value.trim()
  if (!formData.tagList.find((t) => t === val)) {
    formData.tagList.push(newTag.value.trim())
    newTag.value = ''
  }
}

const removeTag = (tag: string) => {
  formData.tagList = formData.tagList.filter((t) => t !== tag)
}

const onSubmit = async () => {
  let article = null
  if (slug.value) {
    article = await putArticle(slug.value, formData)
  } else {
    article = await postArticle(formData)
  }
  if (article) await routerPush('article', { slug: article.slug })
}

async function fetchArticle() {
  if (!slug.value) return
  const article = await getArticleBySlug(slug.value)
  formData.title = article.title
  formData.body = article.body
  formData.description = article.description
  formData.tagList = article.tagList
}

await fetchArticle()
</script>
