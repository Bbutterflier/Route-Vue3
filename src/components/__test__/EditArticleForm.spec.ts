import { createTestingPinia } from '@pinia/testing'
import { flushPromises, mount } from '@vue/test-utils'
import { router, routerPush } from '../../router'
import { getArticleBySlug } from '../../services/article/getArticle'
import { postArticle } from '../../services/article/postArticle'
import asyncComponentWrapper from '../../utils/test/async-component-wrapper'
import fixtures from '../../utils/test/fixtures'
import EditArticleForm from '../EditArticleForm.vue'

jest.mock('src/services/article/postArticle')
jest.mock('src/services/article/getArticle')

describe('EditArticleForm', () => {
  const mockPostArticle = postArticle as jest.MockedFunction<typeof postArticle>
  const mockGetArticle = getArticleBySlug as jest.MockedFunction<
    typeof getArticleBySlug
  >
  beforeEach(async () => {
    await routerPush('create-article', { slug: '' })
  })

  it('should render empty form', async () => {
    const wrapper = mount(asyncComponentWrapper(EditArticleForm), {
      global: {
        plugins: [router, createTestingPinia()],
      },
    })
    await flushPromises()

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render the article values ', async () => {
    await routerPush('article', { slug: 'article-slug' })
    mockGetArticle.mockResolvedValue({ ...fixtures.article })

    const wrapper = mount(asyncComponentWrapper(EditArticleForm), {
      global: {
        plugins: [router, createTestingPinia()],
      },
    })

    await flushPromises()

    expect(
      (wrapper.get('[data-test="title"]').element as HTMLInputElement).value
    ).toBe(fixtures.article.title)

    expect(
      (wrapper.get('[data-test="description"]').element as HTMLInputElement)
        .value
    ).toBe(fixtures.article.description)

    expect(
      (wrapper.get('[data-test="body"]').element as HTMLInputElement).value
    ).toBe(fixtures.article.body)

    expect(wrapper.findAll('.tag-default')).toHaveLength(
      fixtures.article.tagList.length
    )
  })

  test('post a new article', async () => {
    const wrapper = mount(asyncComponentWrapper(EditArticleForm), {
      global: {
        plugins: [router, createTestingPinia()],
      },
    })
    await flushPromises()

    const titleInput = wrapper.get('[data-test="title"]')
    titleInput.setValue('title')

    const descriptionInput = wrapper.get('[data-test="description"]')
    descriptionInput.setValue('description')

    const bodyInput = wrapper.get('[data-test="body"]')
    bodyInput.setValue('body')

    const newTagInput = wrapper.get('[data-test="newTag"]')
    newTagInput.setValue('tag1')
    newTagInput.trigger('keydown.enter')

    await wrapper.get('[type="submit"]').trigger('submit')

    expect(mockPostArticle).toBeCalledWith({
      body: 'body',
      description: 'description',
      tagList: ['tag1'],
      title: 'title',
    })
  })
})
