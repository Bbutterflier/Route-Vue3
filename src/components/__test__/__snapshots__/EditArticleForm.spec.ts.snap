// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditArticleForm should render empty form 1`] = `
"<form>
  <fieldset>
    <fieldset class=\\"form-group\\"><input type=\\"text\\" class=\\"form-control form-control-lg\\" placeholder=\\"Article Title\\" data-test=\\"title\\"></fieldset>
    <fieldset class=\\"form-group\\"><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"What's this article about?\\" data-test=\\"description\\"></fieldset>
    <fieldset class=\\"form-group\\"><textarea class=\\"form-control\\" rows=\\"8\\" placeholder=\\"Write your article (in markdown)\\" data-test=\\"body\\"></textarea></fieldset>
    <fieldset class=\\"form-group\\"><input type=\\"text\\" class=\\"form-control\\" placeholder=\\"Enter tags\\" data-test=\\"newTag\\">
      <div class=\\"tag-list\\" data-test=\\"tag-list\\"></div>
    </fieldset><button class=\\"btn btn-lg pull-xs-right btn-primary\\" type=\\"submit\\"> Publish Article </button>
  </fieldset>
</form>"
`;
